curated_transformers/generation/logits.py

```diff
@@ -0,0 +1,118 @@
+from abc import ABC, abstractmethod
+from collections import UserList
+import torch
+from torch import Tensor
+
+
+class LogitsTransform(ABC):
+    """
+    A logits transform changes the logits of a softmax distribution in some
+    way. For instance, :class:`TemperatureTransform` changes the temperature
+    of the softmax distribution.
+    """
+
+    def __call__(self, logits: Tensor, inplace=False) -> Tensor:
+        """
+        Transform the given logits.
+
+        :param logits:
+            The logits array.
+            **Shape:** (..., n_class)
+        :param inplace:
+            Transform logits in-place.
+        :returns:
+            The transformed logits. This will be the same tensor object as
+            the ``logits`` argument when ``inplace`` is set.
+            **Shape:** (..., n_class)
+        """
+        if not inplace:
+            logits = logits.clone()
+        self._process_logits(logits)
+        return logits
+
+    @abstractmethod
+    def _process_logits(self, logits: Tensor):
+        """
+        Transform logits in-place.
+
+        :param logits:
+            The logits array.
+            **Shape:** (..., n_class)
+        """
+        ...
+
+
+class CompoundLogitTransforms(UserList, LogitsTransform):
+    """
+    Sequentially apply multiple logit transforms.
+    """
+
+    def _process_logits(self, logits: Tensor):
+        for transform in self:
+            logits = transform(logits, inplace=True)
+
+
+class TopKTransform(LogitsTransform):
+    """
+    Set the probability of non-top-k classs to zero. The probability of
+    the classes that are zeroed out is redistributed across the top-k
+    classes.
+    """
+
+    def __init__(self, k: int):
+        """
+        Construct a top-k logits transform.
+
+        :param k:
+            The value of k in top-k. The transform is a noop for values
+            less than 1.
+        """
+        super().__init__()
+        self.k = k
```

Comment:
What speaks against making `k>=1` an requirement?

Reply:
You want some value of `k` where you can effectively disable it. It seems that other implementations use `k<1` to do this. You could set `k` to a large value, bet then you need to know the vocab size.